lib/client/client.go

```diff
@@ -577,58 +577,122 @@ func (client *NodeClient) ExecuteSCP(cmd scp.Command) error {
 	return trace.Wrap(err)
 }
 
-// listenAndForward listens on a given socket and forwards all incoming connections
-// to the given remote address via
-func (client *NodeClient) listenAndForward(socket net.Listener, remoteAddr string) {
-	defer socket.Close()
-	defer client.Close()
-	proxyConnection := func(incoming net.Conn) {
-		defer incoming.Close()
-		var (
-			conn net.Conn
-			err  error
-		)
-		log.Debugf("nodeClient.listenAndForward(%v -> %v) started", incoming.RemoteAddr(), remoteAddr)
-		for attempt := 1; attempt <= 5; attempt++ {
-			conn, err = client.Client.Dial("tcp", remoteAddr)
-			if err != nil {
-				log.Errorf("Connection attempt %v: %v", attempt, err)
-				// failed to establish an outbound connection? try again:
-				time.Sleep(time.Millisecond * time.Duration(100*attempt))
+func (client *NodeClient) proxyConnection(ctx context.Context, incoming net.Conn, remoteAddr string) {
```

Comment:
I find it confusing to have two identical functions with similar names `proxyConn` and `proxyConnection` with only difference being the first one returning error vs. the other one not.

Just use the local closure instead of `proxyConnection` function.